core/src/test/java/kafka/server/SharePartitionManagerTest.java

```diff
@@ -1128,4 +1141,156 @@ public void testShareFetchContextResponseSize() {
     }
+
+    @Test
+    public void testMultipleSequentialShareFetches() {
+
+        String groupId = "grp";
+        Uuid memberId1 = Uuid.randomUuid();
+        FetchParams fetchParams = new FetchParams(ApiKeys.SHARE_FETCH.latestVersion(), FetchRequest.FUTURE_LOCAL_REPLICA_ID, -1, 0,
+                1, 1024 * 1024, FetchIsolation.HIGH_WATERMARK, Optional.empty());
+        Uuid fooId = Uuid.randomUuid();
+        Uuid barId = Uuid.randomUuid();
+        TopicIdPartition tp0 = new TopicIdPartition(fooId, new TopicPartition("foo", 0));
+        TopicIdPartition tp1 = new TopicIdPartition(fooId, new TopicPartition("foo", 1));
+        TopicIdPartition tp2 = new TopicIdPartition(barId, new TopicPartition("bar", 0));
+        TopicIdPartition tp3 = new TopicIdPartition(barId, new TopicPartition("bar", 1));
+        TopicIdPartition tp4 = new TopicIdPartition(fooId, new TopicPartition("foo", 2));
+        TopicIdPartition tp5 = new TopicIdPartition(barId, new TopicPartition("bar", 2));
+        TopicIdPartition tp6 = new TopicIdPartition(fooId, new TopicPartition("foo", 3));
+        Map<TopicIdPartition, Integer> partitionMaxBytes = new HashMap<>();
+        partitionMaxBytes.put(tp0, 40000);
+        partitionMaxBytes.put(tp1, 40000);
+        partitionMaxBytes.put(tp2, 40000);
+        partitionMaxBytes.put(tp3, 40000);
+        partitionMaxBytes.put(tp4, 40000);
+        partitionMaxBytes.put(tp5, 40000);
+        partitionMaxBytes.put(tp6, 40000);
+
+        ReplicaManager replicaManager = Mockito.mock(ReplicaManager.class);
+        SharePartitionManager sharePartitionManager = new SharePartitionManager(replicaManager, new MockTime(),
+                new SharePartitionManager.ShareSessionCache(10, 1000));
+
+        doAnswer(invocation -> {
+            sharePartitionManager.releaseFetchQueueAndPartitionsLock(groupId, partitionMaxBytes.keySet());
+            return null;
+        }).when(replicaManager).fetchMessages(any(), any(), any(ReplicaQuota.class), any());
+
+        sharePartitionManager.fetchMessages(groupId, memberId1.toString(), fetchParams, Arrays.asList(tp0, tp1, tp2, tp3), partitionMaxBytes);
+        Mockito.verify(replicaManager, times(1)).fetchMessages(
+                any(), any(), any(ReplicaQuota.class), any());
+
+        sharePartitionManager.fetchMessages(groupId, memberId1.toString(), fetchParams, Collections.singletonList(tp4), partitionMaxBytes);
+        Mockito.verify(replicaManager, times(2)).fetchMessages(
+                any(), any(), any(ReplicaQuota.class), any());
+
+        sharePartitionManager.fetchMessages(groupId, memberId1.toString(), fetchParams, Arrays.asList(tp5, tp6), partitionMaxBytes);
+        Mockito.verify(replicaManager, times(3)).fetchMessages(
+                any(), any(), any(ReplicaQuota.class), any());
+    }
+
+    @Test
+    public void testMultipleConcurrentShareFetches() throws InterruptedException {
```

Comment:
This is an inherently risky test in terms of reliability. Please make sure you run many, many iterations before merging and that it passes on the CI build. I just want to make sure it's not flaky.

Reply:
Understandable, I have run the test locally over 100 times and it passed everytime. Will make sure it passes on CI build before we merge